### ISSUE TITLE
feat: 集成 NVIDIA Transformer Engine 并按 GPU 架构自动启用多精度支持 (FP16/BF16/FP8/MXFP8/NVFP4)

### ISSUE THREAD

**Issue by Puiching-Memory** [Puiching-Memory/TAAC_2026#28]
## 动机

当前仓库在 `triton_attention.py` 和 `triton_ffn.py` 中手动维护 FP8 Triton 内核，存在以下问题：

- 仅支持 `fp8-e4m3fn` / `fp8-e5m2` 两种精度，无法覆盖 MXFP8（Blackwell block scaling）和 NVFP4（4-bit E2M1）
- 训练时 `requires_grad=True` 会回退到 PyTorch 参考路径（`_can_use_triton_attention` / `_can_use_triton_activation` 中的 guard），FP8 加速仅对推理生效
- scaling factor 管理（delayed scaling / current scaling / block scaling）需要自行实现，维护成本高
- Mask 语义、形状限制（seq_len > 256 / head_dim > 128 回退）等 edge case 需持续人工维护

[NVIDIA Transformer Engine](https://github.com/NVIDIA/TransformerEngine)（TE）已提供成熟的全栈解决方案，覆盖 Linear、LayerNormMLP、DotProductAttention、MultiheadAttention、TransformerLayer 等模块级 FP8/FP4 支持，并被 Megatron-LM、NeMo、DeepSpeed、HuggingFace Accelerate 等主流框架集成。

## 目标

1. **集成 TE 作为可选后端**：在共享算子层 (`transformer.py`) 新增 `"te"` 后端选项
2. **按 GPU 架构自动选择最优精度**：运行时探测 Compute Capability，自动启用该架构支持的最高精度

## GPU 架构 × 精度矩阵

| GPU 架构 | Compute Capability | FP16 | BF16 | FP8 (E4M3/E5M2) | MXFP8 (E4M3 block scaling) | NVFP4 (E2M1) |
|---|---|---|---|---|---|---|
| **Ampere** (A100 等) | SM 8.0 | ✅ | ✅ | ❌ | ❌ | ❌ |
| **Ada** (RTX 4090, L40 等) | SM 8.9 | ✅ | ✅ | ✅ | ❌ | ❌ |
| **Hopper** (H100, H800 等) | SM 9.0 | ✅ | ✅ | ✅ | ❌ | ❌ |
| **Blackwell** (B200 等) | SM 10.0+ | ✅ | ✅ | ✅ | ✅ | ✅ |

> **精度选择策略**：NVFP4 > MXFP8 > FP8 > BF16 > FP16，按硬件上限自动降级；用户可通过配置覆盖。

## TE Recipe 映射

| 精度 | TE Recipe 类 | 关键参数 |
|---|---|---|
| FP8 delayed scaling | `recipe.DelayedScaling` | `fp8_format=Format.HYBRID`, `amax_history_len=16` |
| FP8 current scaling | `recipe.Float8CurrentScaling` | per-tensor just-in-time scaling |
| FP8 block scaling | `recipe.Float8BlockScaling` | DeepSeek-v3 style block scaling |
| MXFP8 | `recipe.MXFP8BlockScaling` | `fp8_format=Format.E4M3`，32-element E8M0 scaling |
| NVFP4 | `recipe.NVFP4BlockScaling` | E2M1 + stochastic rounding + random Hadamard |

### Phase 1: GPU 能力探测与精度路由

- [ ] 增强 `scripts/verify_gpu_env.py`：报告 Compute Capability（当前只报告 `cuda_available` 和 device count）
- [ ] 新建 `src/taac2026/infrastructure/nn/gpu_capability.py`：
  ```python
  def detect_precision_support() -> dict:
      """返回当前 GPU 支持的精度集合及推荐 TE recipe"""
      cap = torch.cuda.get_device_capability()
      major, minor = cap
      result = {"fp16": True, "bf16": major >= 8}
      result["fp8"] = (major > 8) or (major == 8 and minor >= 9)  # Ada 8.9+
      result["mxfp8"] = major >= 10  # Blackwell
      result["nvfp4"] = major >= 10  # Blackwell
      return result
  ```

### Phase 2: TE 后端集成 — 共享算子层

- [ ] 在 `AttentionBackend` 类型中新增 `"te"` 选项（当前为 `"auto" | "torch" | "triton"`）
- [ ] 在 `FeedForwardBackend` 类型中新增 `"te"` 选项
- [ ] `TaacTransformerBlock` 中当 backend=`"te"` 时：
  - Attention → `te.DotProductAttention` 或 `te.MultiheadAttention`
  - FFN → `te.LayerNormMLP`（支持 gelu/silu/swiglu 激活，与现有 `FeedForwardActivation` 对齐）
  - 外层包裹 `te.autocast(enabled=True, recipe=auto_recipe)`

### Phase 3: Mask 语义适配

- [ ] 添加 mask 适配层：本仓库使用 `True=keep`（`scores.masked_fill(~mask, -1e9)`），TE v1.7+ 使用 `True=masked-out`
  ```python
  def adapt_mask_for_te(mask: torch.Tensor) -> torch.Tensor:
      """Invert mask: repo True=keep → TE True=masked-out"""
      return ~mask
  ```

### Phase 4: 依赖与安装

- [ ] `pyproject.toml` 新增可选依赖组：
  ```toml
  [project.optional-dependencies]
  te = ["transformer-engine[pytorch]"]
  ```
- [ ] 安装文档补充 `--no-build-isolation` 说明（TE 要求）和 CUDA/cuDNN 版本要求
- [ ] `verify_gpu_env.py` 中增加 TE 可用性检测

### Phase 5: 逐步覆盖实验包

- [ ] **优先**：使用标准 `TaacTransformerBlock` 的实验包（baseline, interformer, deepcontextnet 等）— 直接受益
- [ ] **其次**：有局部 SwiGLU/FFN 的实验包（uniscaleformer）— FFN 路径可替换
- [ ] **暂不覆盖**：HSTU/SiLU 自定义 attention 路径（hstu.py 中 `SiLUAttention` 使用 `F.silu` 替代 softmax，TE `DotProductAttention` 无法直接替代）

### Phase 6: 测试

- [ ] 单元测试：各精度 recipe 下 TE 模块输出与 PyTorch 参考路径的数值一致性
- [ ] 集成测试：不同 GPU 架构下自动精度选择的正确性（可 mock `torch.cuda.get_device_capability`）
- [ ] 回归测试：确保现有 Triton 路径不受影响（backend=`"triton"` 行为不变）
- [ ] 性能基准：TE 后端 vs 现有 Triton 后端的吞吐量对比

## 不在本 Issue 范围内

- FlashAttention 独立集成（仅 attention，不含 FFN/MLP）— 可作为后续 issue
- torchao / PyTorch AO 集成（training recipes 层级，非模块级替代）
- HSTU `SiLUAttention` 的 TE 适配（需 TE 支持自定义 score function）

## 参考

- [Transformer Engine GitHub](https://github.com/NVIDIA/TransformerEngine)
- [TE FP8/FP4 Primer](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/examples/fp8_primer.html)
- [TE PyTorch API](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/api/pytorch.html)
- [TE Installation Guide](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/installation.html)
- [Pretraining Large Language Models with NVFP4](https://arxiv.org/abs/2509.25149)
- 本仓库相关文件：
  - `src/taac2026/infrastructure/nn/triton_attention.py` — 当前 attention 后端
  - `src/taac2026/infrastructure/nn/triton_ffn.py` — 当前 FFN 后端
  - `src/taac2026/infrastructure/nn/transformer.py` — 共享 Transformer block（集成入口）
  - `src/taac2026/infrastructure/nn/hstu.py` — HSTU 自定义路径（暂不覆盖）
  - `docs/guide/triton-kernels.md` — Triton 策略文档

**Comment by Puiching-Memory** [comment]
**汇报表 1**
| Profile | Scenario | Winner | Torch ms | Triton ms | TE ms | TE/Torch | TE/Triton |
| --- | --- | --- | ---: | ---: | ---: | ---: | ---: |
| hyformer-default | self-masked | triton | 0.465 | 0.419 | 1.722 | 3.71x | 4.11x |
| hyformer-default | self-no-mask | torch | 0.320 | 0.352 | 1.368 | 4.28x | 3.88x |
| hyformer-default | self-causal | triton | 0.457 | 0.392 | 1.401 | 3.06x | 3.58x |
| hyformer-default | cross-masked | torch | 0.382 | 0.401 | 1.872 | 4.90x | 4.66x |
| hyformer-default | cross-no-mask | torch | 0.282 | 0.329 | 1.469 | 5.21x | 4.46x |
| deepcontextnet-default | self-masked | triton | 0.466 | 0.425 | 1.717 | 3.68x | 4.04x |
| deepcontextnet-default | self-no-mask | torch | 0.324 | 0.359 | 1.364 | 4.21x | 3.80x |
| deepcontextnet-default | self-causal | triton | 0.441 | 0.388 | 1.421 | 3.22x | 3.66x |
| deepcontextnet-default | cross-masked | torch | 0.536 | 0.548 | 1.984 | 3.70x | 3.62x |
| deepcontextnet-default | cross-no-mask | triton | 0.331 | 0.329 | 1.453 | 4.39x | 4.42x |
| medium-reference | self-masked | torch | 0.609 | 0.649 | 1.730 | 2.84x | 2.66x |
| medium-reference | self-no-mask | torch | 0.574 | 0.594 | 1.446 | 2.52x | 2.43x |
| medium-reference | self-causal | torch | 0.493 | 0.519 | 1.281 | 2.60x | 2.47x |
| medium-reference | cross-masked | torch | 0.448 | 0.573 | 1.615 | 3.61x | 2.82x |
| medium-reference | cross-no-mask | torch | 0.338 | 0.467 | 1.236 | 3.65x | 2.65x |
| large-reference | self-masked | torch | 1.318 | 1.587 | 1.791 | 1.36x | 1.13x |
| large-reference | self-no-mask | torch | 1.121 | 1.486 | 1.375 | 1.23x | 0.93x |
| large-reference | self-causal | te | 1.295 | 1.518 | 1.280 | 0.99x | 0.84x |
| large-reference | cross-masked | torch | 1.059 | 1.370 | 1.795 | 1.70x | 1.31x |
| large-reference | cross-no-mask | torch | 0.867 | 1.312 | 1.405 | 1.62x | 1.07x |

**汇报表 2**
| Profile | TE self-no-mask vs masked | TE self-causal vs masked | TE cross-no-mask vs masked |
| --- | ---: | ---: | ---: |
| hyformer-default | 20.6% | 18.7% | 21.5% |
| deepcontextnet-default | 20.5% | 17.2% | 26.8% |
| medium-reference | 16.5% | 25.9% | 23.5% |
| large-reference | 23.2% | 28.5% | 21.7% |

**结论**
| 指标 | 结果 |
| --- | --- |
| 总 case 数 | 20 |
| Torch 最快 | 14 |
| Triton 最快 | 5 |
| TE 最快 | 1 |
| TE 唯一领先场景 | large-reference / self-causal |
| 当前判断 | TE fast path 已有效降低开销，但大多数场景仍未超过 torch，优势主要开始出现在 large self-attention 场景 |
